### ISSUE TITLE
Support new relation load strategy for MariaDB (`relationJoins` preview feature)

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#23346]
In versions 5.7.0 and 5.10.0 we released support for a new relation load strategy for PostgreSQL and MySQL as part of the `relationJoins` preview feature: 

- Documentation: https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#relation-load-strategies-preview
- Launch blog post: https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview

Although they usually overlap, this currently does not include MariaDB via the `mysql` datasource provider. 

Unfortunately MariaDB does not support correlated subqueries, the method we use for this new relation load strategy for MySQL. (It also does not supports lateral joins, the method we use for PostgreSQL.) We would need to implement a completely different approach for MariaDB.

This issue is to track demand and use cases for implementing this new relation load strategy for MariaDB.

**Comment by stouch** [comment]
Really expecting this to be resolved. 

Gonna look obvious, but for the moment, as for the issue of "0000-00-00" datetime (null dates in old MariaDB versions), when I need to make JOINs, I create my own queries using $queryRaw : 

```typescript
const myRows = await this.prisma.$queryRaw<{anyColumn: number, otherColumn: string}>`
        SELECT anyColumn, otherColumn,
          FROM yourTable
          JOIN anotherTable ON anotherTable.columnId = yourTable.someColumnId
          WHERE yourCondition = 'something'
`;
```

😅

**Comment by mtabbabi** [comment]
Thanks for all the great work on the relation load strategies feature!

I was wondering if there are any updates regarding potential support for MariaDB?

**Comment by Useems** [comment]
I would really like support for MariaDB. I don't want to have to migrate my production database to MySQL 5.7 and then to 8 just because of Prisma.

**Comment by Useems** [comment]
> Really expecting this to be resolved.
> 
> Gonna look obvious, but for the moment, as for the issue of "0000-00-00" datetime (null dates in old MariaDB versions), when I need to make JOINs, I create my own queries using $queryRaw :
> 
> const myRows = await this.prisma.$queryRaw<{anyColumn: number, otherColumn: string}[]>`
>         SELECT anyColumn, otherColumn
>           FROM yourTable
>           JOIN anotherTable ON anotherTable.columnId = yourTable.someColumnId
>           WHERE yourCondition = 'something'
> `;
> 😅

Yeah.. I needed to create my own queryRaw implementation a few years ago with the ability to replace and cancel the query. Maybe it will help someone.

```ts
/**
* A modification to Prisma's queryRaw that allows you to cancel queries
* In addition to having an implementation similar to Sequelize
* Contains the "replacement" object with properties
*/

// @ts-ignore
import namedPlaceholders from 'named-placeholders';
import prisma from './get-prisma';
import { Sql } from '@prisma/client/runtime/library';
import { Prisma } from '@prisma/client';

const replacement: any = namedPlaceholders();

// @ts-ignore
BigInt.prototype.toJSON = function () {
  return this.toString();
};

export type QueryRawConfig = {
  replacement?: any;
  signal?: any;
  id?: number;
  logging?: (...args: any[]) => void;
};

const queryRaw: (
  query: string,
  config?: QueryRawConfig,
) => Promise<any> = async (query, config = {}) => {
  const id = config.id || parseInt((Math.random() * 10000000).toFixed(0));

  const replica = prisma.$replica();

  var eventFunction = async () => {
    const data: any = await replica.$queryRawUnsafe(
      `SELECT * FROM INFORMATION_SCHEMA.PROCESSLIST WHERE COMMAND = 'Execute' AND INFO LIKE '% WHERE ${id}=${id} AND %';`,
    );

    if (data && data[0]) {
      await replica.$queryRawUnsafe(`KILL ${data[0].ID};`);
    }

    config?.signal?.removeEventListener('abort', eventFunction);
  };

  config?.signal?.addEventListener('abort', eventFunction);

  try {
    if (config.replacement) {
      for (let i in config.replacement) {
        if (typeof config.replacement[i] === 'object') {
          config.replacement[i] =
            config.replacement[i] && config.replacement[i].length > 0
              ? Prisma.join(config.replacement[i])
              : '';
        }
      }
    }

    let newQuery = replacement.parse(
      config.id ? query : query.replace('WHERE', `WHERE ${id}=${id} AND `),
      config.replacement || {},
    )[0];

    let sql = new Sql(
      typeof newQuery === 'string' ? [newQuery] : newQuery,
      replacement(
        config.id ? query : query.replace('WHERE', `WHERE ${id}=${id} AND `),
        config.replacement || {},
      )[1],
    );

    if (config.logging)
      config.logging(
        sql.values
          .map(
            (val, i) =>
              sql.strings[i] +
              (typeof val === 'string' ? '"' + val + '"' : val),
          )
          .join('') + sql.strings[sql.strings.length - 1],
      );

    let data = await replica.$queryRaw(sql);

    return data;
  } catch (err: any) {
    console.log(err);
    if (err) {
      if (err.code === 'P1017' || err.code === 'P1001') {
        return await queryRaw(query, config);
      } else if (
        err.code === 'P2010' &&
        (!err.meta.code || err.meta.code != '1064')
      ) {
        return [];
      } else {
        throw err;
      }
    } else {
      throw err;
    }
  }
};

export default queryRaw;
```

**Comment by barbeque-squared** [comment]
Would it make sense (and be doable) for Prisma to support "normal" joins on MariaDB for only a subset of what `relationJoins` can do on MySQL and Postgres? To make it more concrete:
1. The _vast_ majority of our queries are (_I suspect_ -- see below) ultimately just some (inner) joins with some fixed query parameters.
2. Falling back to the current query behaviour for more complex things is totally acceptable.

The one drawback being that unlike MySQL and Postgres, this would be more "best effort", which might cause some issues if someone explicitly specifies to use `relationJoins` for a specific query, but that query cannot be done that way on MariaDB: does it error out (like it currently does) or does it just fallback to the current query behaviour in that case?

The alternative would be to call this something like `preferRelationJoins` specifically for MariaDB, because "preferring" does not mandate it actually being used.

From a purity perspective, I do agree that it should either implement it for all queries or none. But from a practical perspective, having joins on _some_ queries might be preferable to never having it. What I don't know is what are the specifics that would cause a query to be deemed "too complex -> falling back", or the other way around, the whitelist of query properties (eg 1:1 vs 1:n relations, sorting, limit, model/result extensions, include, etc) that would qualify for being a "simple join" query. Knowing this whitelist would lead to being able to figure out whether statement 1 is true or not, which would in turn (in)validate the usecase.

Parts of this might also be able to be reused for SQL Server and SQLite.
